crates/sui-light-client/src/main.rs

```diff
@@ -0,0 +1,714 @@
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+async fn read_full_checkpoint(checkpoint_path: &PathBuf) -> anyhow::Result<CheckpointData> {
+    let mut reader = fs::File::open(checkpoint_path.clone())?;
+    let metadata = fs::metadata(checkpoint_path)?;
+    let mut buffer = vec![0; metadata.len() as usize];
+    reader.read_exact(&mut buffer)?;
+    bcs::from_bytes(&buffer).map_err(|_| anyhow!("Unable to parse checkpoint file"))
+}
+
+async fn write_full_checkpoint(
+    checkpoint_path: &Path,
+    checkpoint: &CheckpointData,
+) -> anyhow::Result<()> {
+    let mut writer = fs::File::create(checkpoint_path)?;
+    let bytes = bcs::to_bytes(&checkpoint)
+        .map_err(|_| anyhow!("Unable to serialize checkpoint summary"))?;
+    writer.write_all(&bytes)?;
+    Ok(())
+}
+
+async fn get_full_checkpoint(config: &Config, seq: u64) -> anyhow::Result<CheckpointData> {
+    let mut checkpoint_path = config.checkpoint_summary_dir.clone();
+    checkpoint_path.push("untrusted_cache");
+    checkpoint_path.push(format!("{}.bcs", seq));
+
+    // Try reading the cache
+    if let Ok(checkpoint) = read_full_checkpoint(&checkpoint_path).await {
+        return Ok(checkpoint);
+    }
+
+    // Downloading the checkpoint from the server
+    let client: Client = Client::new(config.rest_url());
+    let full_check_point = client.get_full_checkpoint(seq).await?;
+
+    // Add to cache
+    write_full_checkpoint(&checkpoint_path, &full_check_point).await?;
+
+    Ok(full_check_point)
+}
+
+fn extract_verified_effects_and_events(
+    checkpoint: &CheckpointData,
+    committee: &Committee,
+    tid: TransactionDigest,
+) -> anyhow::Result<(TransactionEffects, Option<TransactionEvents>)> {
+    let summary = &checkpoint.checkpoint_summary;
+
+    // Verify the checkpoint summary using the committee
+    summary.clone().verify(committee)?;
+
+    // Check the validity of the checkpoint contents
+    let contents = &checkpoint.checkpoint_contents;
+    anyhow::ensure!(
+        contents.digest() == &summary.content_digest,
+        "The content digest in the checkpoint summary does not match the digest of the checkpoint contents"
+    );
+
+    // Check the validity of the transaction
+    let (matching_tx, _) = checkpoint
+        .transactions
+        .iter()
+        .zip(contents.iter())
+        // Note that we get the digest of the effects to ensure this is
+        // indeed the correct effects that are authenticated in the contents.
+        .find(|(tx, digest)| {
+            tx.effects.execution_digests() == **digest && digest.transaction == tid
+        })
+        .ok_or(anyhow!("Transaction not found in checkpoint contents"))?;
+
+    // Check the events are all correct.
+    let events_digest = matching_tx.events.as_ref().map(|events| events.digest());
+    anyhow::ensure!(
+        events_digest.as_ref() == matching_tx.effects.events_digest(),
+        "Events digest does not match"
+    );
```

Comment:
Some of the verifications about CheckpointData probably should have been verified right after downloading. That way 1) we have trusted data in the cache 2) it's easier to reason about the code because we know what to trust at the beginning of each function. The current structure of the code makes that a bit hard to think what has been verified and what hasn't.

Reply:
Removing the cache to avoid this issue for now.